idea/idea-fir/tests/org/jetbrains/kotlin/checkers/AbstractFirKotlinHighlightingPassTest.kt

```diff
@@ -5,23 +5,76 @@
 
 package org.jetbrains.kotlin.checkers
 
+import com.intellij.openapi.util.io.FileUtil
 import com.intellij.rt.execution.junit.FileComparisonFailure
+import junit.framework.Assert
 import org.jetbrains.kotlin.idea.invalidateCaches
 import org.jetbrains.kotlin.idea.test.withCustomCompilerOptions
 import org.jetbrains.kotlin.idea.withPossiblyDisabledDuplicatedFirSourceElementsException
 import org.jetbrains.kotlin.psi.KtFile
 import org.jetbrains.kotlin.test.uitls.IgnoreTests
+import org.jetbrains.kotlin.test.uitls.IgnoreTests.DIRECTIVES
 import java.io.File
 
 abstract class AbstractFirKotlinHighlightingPassTest : AbstractKotlinHighlightingPassTest() {
+    companion object {
+        private val isTeamCityBuild: Boolean = System.getProperty("TEAMCITY_VERSION") != null
+    }
+
     override val captureExceptions: Boolean = false
 
     override fun isFirPlugin(): Boolean = true
 
+    override fun fileName(): String {
+        val originalTestFile = originalTestFile()
+        if (originalTestFile.readText().startsWith(DIRECTIVES.FIR_IDENTICAL)) {
+            return originalTestFile.name
+        }
+        val firTestFile = firTestFile()
+        if (!firTestFile.exists()) {
+            FileUtil.copy(originalTestFile, firTestFile)
+        }
+        return firTestFile.name
+    }
+
+    private fun originalTestFile(): File {
+        return File(testPath(super.fileName()))
+    }
+
+    private fun firTestFile(): File {
+        val originalTestFileName: String = super.fileName()
+        return File(testPath(originalTestFileName.removeSuffix(".kt") + ".fir.kt"))
+    }
+
     override fun doTest(filePath: String) {
-        IgnoreTests.runTestIfEnabledByFileDirective(testDataFilePath(), IgnoreTests.DIRECTIVES.FIR_COMPARISON) {
+        IgnoreTests.runTestIfNotDisabledByFileDirective(originalTestFile().toPath(), DIRECTIVES.IGNORE_FIR) {
             myFixture.configureByFile(fileName())
             checkHighlighting(checkWarnings = false, checkInfos = false, checkWeakWarnings = false)
+            val firTestFile = firTestFile()
+            val originalTestFile = originalTestFile()
+            if (firTestFile.exists() && firTestFile.readText().trim() == originalTestFile.readText().trim()) {
```

Comment:
Let's extract all FIR_IDENTICAL related logic to `IgnoreTests`, so it can be reused from other test if needed

Reply:
Sounds good!